Amend / Converge on Slep 6 on sample props aka now as metadata (#50)
scikit-learn/enhancement_proposals: slep006/proposal.rst

    ...          clf__sample_weight=[.5, .7])  # doctest: +SKIP

Several other meta-estimators, such as GridSearchCV, support forwarding these
fit parameters to their base estimator when fitting. Yet a number of important
use cases are currently not supported.

> Features we currently do not support and which to include:

Reviewer: I don't understand the new version here

Reviewer 2: ```suggestion
Features we currently do not support and wish to include:
```

* passing sample properties (e.g. `sample_weight`) to a scorer used in
  cross-validation
* passing sample properties (e.g. `groups`) to a CV splitter in nested cross
  validation
* passing sample properties (e.g. `sample_weight`) to some
  scorers and not others in a multi-metric cross-validation setup
* passing sample properties to non-fit methods, for
  instance to index grouped samples that are to be treated as a single sequence
  in prediction.

The last two items are considered in the API design, yet will not be considered
in the first version of the implementation.

Naming
------

"Sample props" has become a name understood internally to the Scikit-learn
development team. For ongoing usage we propose to use as naming ``metadata``.

Definitions
-----------

consumer
    An estimator, scorer, splitter, etc., that receives and can make use of
    one or more passed metadata.
key
    A label passed along with sample metadata to indicate how it should be
    interpreted (e.g. "weight").

> .. XXX : remove router?

Reviewer: I think it's a useful concept, but if it's not used below, then okay.

* we can consider the use of keys that are not limited to strings or valid
  identifiers (and hence are not limited to using ``_`` as a delimiter).

Naming
------

"Sample props" has become a name understood internally to the Scikit-learn

> development team. For ongoing usage we propose to use as naming ``metadata``.

Reviewer 2: This proposal doesn't add anything new to what we support now. Right now users can already pass column aligned or random metadata in a pipeline, this just keeps that behavior. We wouldn't be able to easily move to the new behavior if we don't support the old behavior.

And I don't mind having non-sample aligned data anyway, it's useful :P

  it. This implies that `clone` needs to be extended to explicitly copy
  request information.

Proposed public syntax:

* `BaseEstimator` will have a method `get_metadata_request`
* Estimators that can consume `sample_weight` will have a `request_sample_weight`
  method available via a mixin.
* `make_scorer` will have a `request_metadata` parameter to specify the requested
  metadata by the scorer.
* `get_metadata_request` will return a dict, whose keys are names of estimator
  methods (`fit`, `predict`, `transform` or `inverse_transform`) and values are
  dictionaries. These dictionaries map the input parameter names to requested
  metadata keys. Example:

  >>> estimator.get_metadata_request()
  {'fit': {'my_sample_weight': {'sample_weight'}}, 'predict': {}, 'transform': {},
  'score': {}, 'split': {}, 'inverse_transform': {}}

* Methods like `request_sample_weight` will have a signature such as:
  `request_sample_weight(self, *, fit=None, score=None)` where fit keyword
  parameter can be `None`, `True`, `False` or a `str`. `str` allows here
  to request a metadata whose name is different from the keyword parameter.

>   Here ``None`` is a default, and ``False`` has a different semantic which

Reviewer: ```suggestion
  Here ``None`` is a default, and ``False`` has a different meaning, which
```
I'm struggling with "semantic" used as a singular noun.

TODO: case involving score() method, e.g. not specifying scoring in
cross_val_score when wrapping an estimator with weighted score func ...

> .. Adrin please motivate this case D

Reviewer 2: This is hard to motivate with what we have inside sklearn. But generally, you can have `groups` used in a CV, which contains _batches_ of data as groups, and then an estimator which takes `groups` as _sensitive attributes_ to a fairness related model.

Or just two third party libraries' estimators having the same name for a parameter, but with completely different semantics.

Solution sketches will import these definitions:

.. literalinclude:: defs.py

The following solution has emerged as the way to move forward,
yet others where considered. See :ref:`slep_006_other`.

Solution: Each child requests
-----------------------------

.. note::

  This solution was known as solution 4 during the discussions.

A meta-estimator provides only what each of its children requests.

> A meta-estimator would also need to request, on behalf of its children,
> any prop that descendant consumers require. 

Reviewer 3: ```suggestion
A meta-estimator would also need to request, on behalf of its children,
any metadata that descendant consumers require.
```
Can we be more precise with what a *descendant consumer* is? Like a pipeline (the meta-estimator) in a pipeline (descendant consumer)?

Reviewer 2: How would you be more specific? meta-estimators are very varied in the way they wrap other estimators

Reviewer 3: Is "decentant consumers" = "its children" or is "descendant consumers" = "caller of this meta-estimator"?

Reviewer 2: oh, my understanding is "its children".

Reviewer: Hmm... "descendant" is a reference to the tree structure of a composite/meta estimator, but that is not made explicit. It might be worth noting that in the definitons above.

Reviewer 3: Thanks for clarifying.

Each object that could receive metadata should have a method called

> `get_metadata_request()` which returns a dict that specify which

Reviewer 3: ```suggestion
`get_metadata_request()` which returns a dict that specifies which
```

> metadata is consumed by each of its methods (keys are of this dictionary

Reviewer 3: ```suggestion
metadata is consumed by each of its methods (keys of this dictionary
```

Possible public syntax:

* `BaseEstimator` will have a method `get_metadata_request`
* Estimators that can consume `sample_weights` will have a `request_sample_weight`
  method available via a mixin.
* `make_scorer` will have a `request_props` parameter to set props required by
  the scorer.
* `get_metadata_request` will return a dict. It maps the key that the user
  passes to the key that the estimator expects.

> .. XXX should be amended

Reviewer 2: in what way?

  details of the solution.
* Props will be known simply as Metadata.
* `**kw` syntax will be used to pass props by key.
* Default requests: A group splitter should return `['groups']` by default
  because requiring `groups` is essential to their functionality. In other
  cases, as for `sample_weight` where an estimator or scorer can operate
  without it, no default request should be assumed, and the consumer should

>   have an explicit request for the prop. In the absence of an explicit request,

Reviewer 3: ```suggestion
  make an explicit request for that metadata. In the absence of an explicit request,
```

>   or explicit request to disregard a prop: if a consumer is capable of
>   requesting some prop by key `p`, and a prop called `p` is passed by the user,

Reviewer 3: ```suggestion
  or explicit request to disregard a specific metadata: if a consumer is capable of
  requesting some metadata by key `p`, and metadata called `p` is passed by the user,
```
